python/python/lance/arrow.py

```diff
@@ -11,14 +11,182 @@
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_uris(cls, uris: Union[pa.StringArray, List[str]]):
+        if isinstance(uris, list):
+            return cls.from_storage(ImageURIType(), pa.array(uris, type=pa.string()))
+        else:
+            return cls.from_storage(ImageURIType(), uris)
+
+    def materialize(self):
+        # Assume first URI's filesystem is the same for all URIs
+        from pyarrow import fs
+
+        first_uri = self.storage[0].as_py()
+        filesystem, _ = fs.FileSystem.from_uri(first_uri)
+
+        # Read all URIs and return their content as lance.arrow.EncodedImageArray
+        images = []
+        for uri in self.storage:
+            with filesystem.open_input_stream(uri.as_py()) as f:
+                images.append(f.read())
+
+        return EncodedImageArray.from_binary_array(pa.array(images, type=pa.binary()))
+
+
+class EncodedImageArray(pa.ExtensionArray):
+    def __repr__(self):
+        return "<lance.arrow.EncodedImageArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_binary_array(cls, array: pa.binary()):
+        return cls.from_storage(EncodedImageType(), array)
+
+    def materialize(self):
+        # TODO: try other decoders (pillow/torchvision/cv2/scipy/..) to decode images
+        # or rather just pick one and make it a requirement?
+        import numpy as np
+
+        try:
+            import tensorflow as tf
+
+            image_arrays = (tf.io.decode_image(img.as_py()) for img in self.storage)
+        except ImportError:
+            import io
+
+            from PIL import Image
+
+            image_arrays = (
+                np.array(Image.open(io.BytesIO(img.as_py()))) for img in self.storage
+            )
+        except ImportError:
+            raise "No image decoder available"
+
+        image_array = np.stack(image_arrays)
+        return ImageTensorArray.from_numpy_ndarray(image_array)
+
+
+class ImageTensorArray(pa.ExtensionArray):
+    import numpy as np
+
+    def __repr__(self):
+        return "<lance.arrow.ImageTensorArray object at 0x%016x>\n%s" % (
+            id(self),
+            repr(self.to_pylist()),
+        )
+
+    @classmethod
+    def from_numpy_ndarray(cls, array: np.ndarray):
+        arrow_type = pa.from_numpy_dtype(array.dtype)
+        shape = array.shape[1:]
+        size = array.size / array.shape[0]
+
+        tensor_array = pa.FixedShapeTensorArray.from_storage(
+            pa.fixed_shape_tensor(arrow_type, shape),
+            pa.FixedSizeListArray.from_arrays(array.ravel(), size),
+        )
+        return cls.from_storage(ImageTensorType(arrow_type, shape), tensor_array)
+
+    def to_tf(self):
+        import tensorflow as tf
+
+        return tf.convert_to_tensor(self.storage.to_numpy_ndarray())
+
+    def to_tfrecords(self):
+        # TODO: write to tfrecords
+        pass
```

Comment:
I think we can skip this. Once converted to TF tensors, users can write them to tfrecords themselves.

Reply:
Removing.